Wardrobe - New: System to modify current wearables (#10606)
acemod/ACE3: docs/wiki/framework/wardrobe-framework.md

---
layout: wiki
title: Wardrobe Framework
description: Explains the wardrobe system and how to make a compat.
group: framework
order: 5
parent: wiki
mod: ace
version:
  major: 3
  minor: 18
  patch: 2
---

## 1. Overview

## 2. Config

> - Only directly defined Subclasses will be taken into account. An fully inherited subclass will be ignored and will not provide an Interaction.

Reviewer: ```suggestion
- Only directly defined Subclasses will be taken into account. A fully inherited subclass will be ignored and will not provide an Interaction.
```

| 0  | Unique Identifier | String | Required |
| 1  | Display Name | String | Required |
| 2  | Required Item | String | Required |
| 3  | Textures | Array | Required |
| 4  | Icon | String | Optional (default: `""` - Default white point) |
| 5  | Material Paths | Array | Optional (default: `[]` - No custom material) |
| 6  | Tag Model | String | Optional (default: `"UserTexture1m_F"` - 1x1m texture surface) |
| **R** | Successfully Added Tag | Boolean | Return value |

#### 2.1.1 Example

`["ace_victoryRed", "Victory Red", "ACE_SpraypaintRed", ["tagTexture1.paa", "tagTexture2.paa"], "icon.paa"] call ace_tagging_fnc_addCustomTag;`

|    | Arguments | Explanation |
| ---| --------- | ----------- |
| 0  | `"ace_victoryRed"` | Unique identifier (similar to class name) |
| 1  | `"Victory Red"` | Name of your tag being displayed in the interaction menu |
| 2  | `"ACE_SpraypaintRed"` | Required item to have in the inventory to be able to spray your tag |
| 3  | `["tagTexture1.paa", "tagTexture2.paa"]` | List of texture variants (one is randomly selected when tagging) |
| 4  | `"icon.paa"` | Icon being displayed in the interaction menu |

### 2.2 Tags in description.ext

> Tags can also be configured in description.ext like shown above.
> File Paths will be relative to your mission, if you want to define Tags inside description.ext but use Addon paths for `Texture`/`Material`/`TagModel` you need to prefix the path with `@`.

Reviewer: ```suggestion
Tags can also be configured in `description.ext` like shown above.
File Paths will be relative to your mission, if you want to define Tags inside `description.ext` but use Addon paths for `Texture`/`Material`/`TagModel` you need to prefix the path with `@`.
```

Author: i hope you are aware that this part will be deleted later, yes?

Reviewer: Totally possible, my mind is easily distracted to forget things, especially when trying to multitask.

---
layout: wiki
title: Wardrobe Framework
description: Explains the wardrobe system and how to make a compat.
group: framework
order: 5
parent: wiki
mod: ace
version:
  major: 3
  minor: 20
  patch: 0
---

## 1. Overview

The Wardrobe addon gives the player the opportunity to change/modify their current wearables into different variants.

> For example, if a uniform has a normal variant and a "rolled-Up sleeves" variant, the player will be able to use ether an ACE Self Interaction or the CBA Context Menu in the inventory.

Reviewer 2: ```suggestion
For example, if a uniform has a normal variant and a "rolled-up sleeves" variant, the player will be able to use ether an ACE self-interaction.
```

Since there is no reliable, common patterns in terms of class inheritance, not even within the same DLC, each possible variant has to be defined individually within `configFile >> "ace_wardrobe"`.

### 1.1 Components

Components are an optional assistance to define a group of variants where a variant is made of one or more components.
For example, a bandana with sunglasses would have the the bandana and the sunglasses as components, while the bandana and sunglasses only have themselves as a component.

> If the player switches from a variant with more components to a variant with less components, the surplus components will be deposited in their inventory. Vice Versa, if the player wants to change from a variant with less components to a variant with more components, they are required to have the missing components in their inventory, which will be removed on conversion.

Reviewer 2: 'Vice Versa' at the beginning of a sentence is grammatically correct, but it sounds weird. Suitable replacements, imo, are 'conversely' and less formal 'on the other hand'.

Author: 08d0f11f5dec85f455df46d076dc15f275c10f2a

        class modifiableTo {
            class G_Bandanna_blk;
            class G_Aviator;
        };
        components[] = { "G_Aviator", "G_Bandanna_blk" };
    };
};
```

## 5. Sounds
The following `CfgSounds` classes are integrated in `ace_wardrobe` and are the default sounds for the `ace_wardrobe_base` and `ace_wardrobe_base_H_visor_up`/`ace_wardrobe_base_H_visor_down` base classes.

### 5.1 Integrated Sounds
The number at the end of the classnames indicates the length of the file in 1/10th seconds.
10 -> 1 sec, 15 -> 1.5sec, ...

- `ace_wardrobe_fabric_06`
- `ace_wardrobe_fabric_07`
- `ace_wardrobe_fabric_16`
- `ace_wardrobe_fabric_20`
- `ace_wardrobe_fabric_25`
- `ace_wardrobe_helmet_visor_05`

## 6. Compatibility

## 6.1 MagzineID
Currently, `ace_intelitems` and `ace_overheating` (spare barrels) are being directly supported.

If an addon or mod utilizes a magazine's `magazineID` to handle additional data about items carried by the player, then the process of modifying a wearable container (uniform, vest, backpack) to another variant will result in new `magazineID`s for all magazines on the player and therefore, require special handling within `ace_wardrobe` functions.

> There *might* be some changes coming from BI regarding how arma handles the `setUnitLoadout` which could resolve the need for this.
> 

Reviewer 2: This should reflect facts only, no speculation.

```suggestion
```

---
layout: wiki
title: Wardrobe Framework
description: Explains the wardrobe system and how to make a compat.
group: framework
order: 5
parent: wiki
mod: ace
version:
  major: 3
  minor: 20
  patch: 0
---

## 1. Overview

The Wardrobe Addon gives the player the opportunity to change/modify their current wearables into different variants.

For example, if a uniform has a normal variant and a "Rolled-Up Sleeves" variant, the player will be able to use ether an ACE Self Interaction or the CBA Context Menu in the Inventory.

Since there is no reliable, common patterns in terms of class inheritance, not even within the same DLC, each possible variant has to be defined individually within `configFile >> "ace_wardrobe"`.

### 1.1 Components

Components are an optional assistance to define a group of variants where a variant is made of one or more components.

> For example, a bandanna with sunglasses would have the the bandanna and the sunglasses as components, while the bandanna and sunglasses themself only have themselves as a component.

Author: since the classnames are written like this `G_Bandanna_blk` and wiki and oxford dict both list both versions as possible variants, i would stick with `bandanna` unless its really required to change it.

https://www.wikiwand.com/en/articles/Bandana
https://www.oxfordlearnersdictionaries.com/definition/english/bandana?q=bandanna

> For example, a bandanna with sunglasses would have the the bandanna and the sunglasses as components, while the bandanna and sunglasses themself only have themselves as a component.

Reviewer 2: Display name has 1 n, so use that.

More examples can be found below.

### 1.2 Config Guidelines

- A Container (Uniform, Vest, Backpack) should never change its `maxLoad`, unless there is a really good reason for it. This will result in the loss of items.
- Variants should not "magically" add/remove parts to/from themselves, unless it's handled through a component, especially when they add protection or functionality.
  - Do: Uniforms with and without Gloves should be fine in most cases, as they are cosmetical and don't break continuity/logic.
  - Do: Helmets with integral flip-down visors.
  - Don't: Helmet with additional armor or face protector without the requirement of some component that will be removed from the inventory.
- There should not be a difference in weight or container capacity between the different variants unless components get added/removed.
  - For Example: `Mass of Helmet with Goggles = Mass of Helmet + Mass of Goggles.`

## 2. Config

> Only directly defined Subclasses will be taken into account. A fully inherited subclass will be ignored and will not provide any interaction. The `ace_wardrobe` properties are to be found as a subclass of the item itself. Multiple base configs are designed at root of `configFile`.

Reviewer 2: I have noticed a lot of capitalised words that shouldn't be. I understand in some instances, but imo it's not warranted in most cases I've seen so far.
```suggestion
Only directly defined subclasses will be taken into account. A fully inherited subclass will be ignored and will not provide any interaction. The `ace_wardrobe` properties are to be found as a subclass of the item itself. Multiple base configs are designed at root of `configFile`.
```

---
layout: wiki
title: Wardrobe Framework
description: Explains the wardrobe system and how to make a compat.
group: framework
order: 5
parent: wiki
mod: ace
version:
  major: 3
  minor: 20
  patch: 0
---

## 1. Overview

The Wardrobe Addon gives the player the opportunity to change/modify their current wearables into different variants.

For example, if a uniform has a normal variant and a "Rolled-Up Sleeves" variant, the player will be able to use ether an ACE Self Interaction or the CBA Context Menu in the Inventory.

> Since there is no reliable, common pattern in terms of class inheritence, not even within the same DLC, each possbile variant has to be defined individually within the classes config properties.

Reviewer: ```suggestion
Since there is no reliable, common pattern in terms of class inheritance, not even within the same DLC, each possible variant has to be defined individually within the classes' config properties.
```

### 1.1 Components

Components are an optional assistance to define a group of variants where a variant is made of one or more components.
For example, a bandanna with sunglasses would have the the bandanna and the sunglasses as components, while the bandanna and sunglasses themself only have themselves as a component.

If the player switches from a variant with more components to a variant with less components, the surplus components will be deposited in their inventory. Vise Versa, if the player wants to change from a variant with less components to a variant with more components, they are required to have the missing components in their inventory, which will be removed on conversion.

More examples can be found below.

### 1.2 Config Guidelines

> - An Container (Uniform, Vest, Backpack) should never change its maxLoad, unless there is a really good reason for it. This will result in the loss of items.

Reviewer: ```suggestion
- A Container (Uniform, Vest, Backpack) should never change its `maxLoad`, unless there is a really good reason for it. This will result in the loss of items.
```

- Variants should never "magically" add/remove parts of themselves, unless its handled through a component.
- There should not be a difference in weight between the different variants unless compoents get added/removed.
  - For Example: `Mass of Helmet with Goggles = Mass of Helmet + Mass of Goggles.`

## 2. Config

> Only directly defined Subclasses will be taken into account. A fully inherited subclass will be ignored and will not provide any interaction. The ace_wardrobe properties are to be found as a subclass of the item itself. Multiple base configs are designed at root of `configFile`.

Reviewer: ```suggestion
Only directly defined Subclasses will be taken into account. A fully inherited subclass will be ignored and will not provide any interaction. The `ace_wardrobe` properties are to be found as a subclass of the item itself. Multiple base configs are designed at root of `configFile`.
```

## 2.1 Properties

| Class Property |  Data Type | Description |
| -------------- |  ----------- | ----------- |

> | modifiableTo[] | Array of Classnames | possible Variants this item can be turned into. |
> | components[] | Array of Classnames  | Components the current variant contains within itself | 
> | sound[]  | Array of CfgSound Entries | to be chosen by random when the action is performed |
> | sound_timing | Number 0..1 | defines the point time relative to the duration when the sound is played |
> | gesture | String of Classname | Gesture to be played when the action is performed |
> | alternativePicture | String of path to icon | to be used instead of target variant picture |
> | alternativeDisplayName | String | Will be used instead of the target variants displayname |
> | duration | Number in seconds | Duration of action. Items are being replaced at the end. |

Reviewer: ```suggestion
| `modifiableTo[]` | Array of Classnames | possible Variants this item can be turned into. |
| `components[]` | Array of Classnames  | Components the current variant contains within itself | 
| `sound[]` | Array of CfgSound Entries | to be chosen by random when the action is performed |
| `sound_timing` | Number 0..1 | defines the point time relative to the duration when the sound is played |
| `gesture` | String of Classname | Gesture to be played when the action is performed |
| `alternativePicture` | String of path to icon | to be used instead of target variant picture |
| `alternativeDisplayName` | String | Will be used instead of the target variants displayname |
| `duration` | Number in seconds | Duration of action. Items are being replaced at the end. |
```

    alternativeDisplayName = CSTRING(jacketClose);
};
```

#### 2.2.5 Base Helmet Visor Up / Down
Common Base Class for Helmets with a Visor that can be flipped up or down.
Here, the duration is carefully timed to be aligned with the "click" of the soun.

```cpp
class ace_wardrobe_base_H_visor_up: ace_wardrobe_base {
    duration = 0.25;
    gesture ="gestureNod";

    sound[] = {CN_SOUND(helmet_visor,05)};
    sound_timing = 0;

    alternativeDisplayName = CSTRING(visorUp);
};

class ace_wardrobe_base_H_visor_down: ace_wardrobe_base {
    duration = 0.25;
    gesture ="gestureNod";

    sound[] = {"ace_wardrobe_05"};
    sound_timing = 0;

    alternativeDisplayName = CSTRING(visorDown);
};
```
#### 2.2.6 Base Headgear with Goggles on / off
Common Base Class for Headgear with goggles that can be used as a facewear item.
```cpp
class ace_wardrobe_base_H_goggles_on: ace_wardrobe_base {
    gesture ="GestureWipeFace";
    alternativeDisplayName = CSTRING(gogglesOn);
};

class ace_wardrobe_base_H_goggles_off: ace_wardrobe_base {
    gesture ="GestureWipeFace";
    alternativeDisplayName = CSTRING(gogglesOff);
};
```

## 3. Examples

### 3.1 Linear Example

> - first, we import the wardrobe base classes `ace_wardrobe_base_U_sleeves_down` and `ace_wardrobe_base_U_sleeves_up` at the root of `configFile`.

Reviewer: ```suggestion
- First, we import the wardrobe base classes `ace_wardrobe_base_U_sleeves_down` and `ace_wardrobe_base_U_sleeves_up` at the root of `configFile`.
```

> - The uniform we would like to configure is part of CfgWeapons and inherits from `Uniform_Base`.

Reviewer: ```suggestion
- The uniform we would like to configure is part of `CfgWeapons` and inherits from `Uniform_Base`.
```

> - Then we edit the desired classes and create the `ace_wardrobe` subclass by inheriting from the partent base classes.

Reviewer: ```suggestion
- Then we edit the desired classes and create the `ace_wardrobe` subclass by inheriting from the parent base classes.
```

- Afterwards, we define which variants this class can be modified to.

> - This Example does not require the use of components since we're not adding or removing anything.

Reviewer: ```suggestion
- This example does not require the use of components since we're not adding or removing anything.
```

```cpp
// config.cpp
class ace_wardrobe_base_U_sleeves_down;
class ace_wardrobe_base_U_sleeves_up;

class CfgWeapons {
    class Uniform_Base;

    // Sleeves Down Variant
    class U_B_CTRG_1: Uniform_Base {
        class ace_wardrobe: ace_wardrobe_base_U_sleeves_down {
            modifiableTo[] = { "U_B_CTRG_3" };
            components[] = {};
        };
    };

    // Sleeves Up Variant
    class U_B_CTRG_3: Uniform_Base {
        class ace_wardrobe: ace_wardrobe_base_U_sleeves_up {
            modifiableTo[] = { "U_B_CTRG_1" };
            components[] = {};
        };
    };
};
```

> ### 3.2 Complex Example with partial use of components

Reviewer: ```suggestion
### 3.2 Complex example with partial use of components
```

```cpp
// config.cpp
class ace_wardrobe_base;

class CfgWeapons {
    // BASECLASS
    class H_Shemag_khk;
    

>     // WS Turban with Balistic Mask

Reviewer: ```suggestion
    // WS Turban with Ballistic Mask
```

    class H_turban_02_mask_black_lxws: H_Shemag_khk {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "lxWS_H_turban_01_black", "lxWS_H_turban_02_black", "lxWS_H_turban_03_black", "lxWS_H_turban_04_black" };
            components[] = { "lxWS_H_bmask_base" };
        };
    };
    // WS Turban Variant 1
    class lxWS_H_turban_01_black: H_Shemag_khk {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "lxWS_H_turban_02_black", "lxWS_H_turban_03_black", "lxWS_H_turban_04_black", "H_turban_02_mask_black_lxws" };

>             components[] = { }; // not required to be defined since it empty array get inherited from the base parent.

Reviewer: Not sure what the wording of the comment here and later is meant to be.

Author: rephrased in most recent commit

>             components[] = { }; // not needed when empty - only shown as an examlpe for how/when to use components

Reviewer 3: ```suggestion
            components[] = { }; // not needed when empty - only shown as an example for how/when to use components
```

        };
    };
    // WS Turban Variant 2
    class lxWS_H_turban_02_black: lxWS_H_turban_01_black {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "lxWS_H_turban_01_black", "lxWS_H_turban_03_black", "lxWS_H_turban_04_black", "H_turban_02_mask_black_lxws" };

>             components[] = { }; // not needed when empty - only shown as an examlpe for how/when to use components

Reviewer 3: ```suggestion
            components[] = { }; // not needed when empty - only shown as an example for how/when to use components
```

        };
    };
    // WS Turban Variant 3
    class lxWS_H_turban_03_black: lxWS_H_turban_01_black {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "lxWS_H_turban_01_black", "lxWS_H_turban_02_black", "lxWS_H_turban_04_black", "H_turban_02_mask_black_lxws" };

>             components[] = { }; // not needed when empty - only shown as an examlpe for how/when to use components

Reviewer 3: ```suggestion
            components[] = { }; // not needed when empty - only shown as an example for how/when to use components
```

        };
    };
    // WS Turban Variant 4
    class lxWS_H_turban_04_black: lxWS_H_turban_01_black {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "lxWS_H_turban_01_black", "lxWS_H_turban_02_black", "lxWS_H_turban_03_black", "H_turban_02_mask_black_lxws" };

>             components[] = { }; // not needed when empty - only shown as an examlpe for how/when to use components

Reviewer 3: ```suggestion
            components[] = { }; // not needed when empty - only shown as an example for how/when to use components
```

        };
    };
};

```

### 3.3 Complex Example with thorough use of components

```cpp
// config.cpp
class ace_wardrobe_base;

class CfgGlasses {
    class None;
    class G_Balaclava_blk;
    class G_Bandanna_shades;

    // Aviator Sunglasses
    class G_Aviator: None {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "G_Bandanna_aviator" };
            components[] = {"G_Aviator"};
        };
    };
    // Bandana, Black with Aviator Sunglasses
    class G_Bandanna_aviator: G_Bandanna_shades {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "G_Bandanna_blk", "G_Aviator" };       
            components[] = { "G_Bandanna_blk", "G_Aviator" };
        };
    };
    // Bandana, Black
    class G_Bandanna_blk: G_Balaclava_blk {
        class ace_wardrobe: ace_wardrobe_base {
            modifiableTo[] = { "G_Bandanna_aviator" };
            components[] = { "G_Bandanna_blk" };
        };
    };
};
```

## 4. Addon Settings

## 5. Sounds

> The following CfgSounds classes are integrated in ace_wardrobe and are the default sounds for the `ace_wardrobe_base` and `ace_wardrobe_base_H_visor_up`/`ace_wardrobe_base_H_visor_down` base classes.

Reviewer: ```suggestion
The following `CfgSounds` classes are integrated in `ace_wardrobe` and are the default sounds for the `ace_wardrobe_base` and `ace_wardrobe_base_H_visor_up`/`ace_wardrobe_base_H_visor_down` base classes.
```

### 5.1 Integrated Sounds

The number at the end of the classnames indicates the length of the file in 1/10th seconds.
10 -> 1 sec, 15 -> 1.5sec, ...

- `ace_wardrobe_fabric_06`
- `ace_wardrobe_fabric_07`
- `ace_wardrobe_fabric_16`
- `ace_wardrobe_fabric_20`
- `ace_wardrobe_fabric_25`
- `ace_wardrobe_helmet_visor_05`

## 6. Compatibility
Currently, ace_IntelItems and ace_overheating (spare barrels) are being directly supported.

> If an addon or mod utilizes a magazines magazineID to handle additional data about an items carried by the player, then the process of modifying a wearable container (uniform, vest, backpack) to another variant will result in new magazineIDs for all magazines on the player.

Reviewer: ```suggestion
If an addon or mod utilizes a magazine's `magazineID` to handle additional data about items carried by the player, then the process of modifying a wearable container (uniform, vest, backpack) to another variant will result in new `magazineID`s for all magazines on the player.
```
